language/bytecode_verifier/src/instantiation_loops.rs

```diff
@@ -0,0 +1,202 @@
+                        let node_to = self.get_or_add_node(Node(*other_func_idx, formal_idx));
+                        if let SignatureToken::TypeParameter(actual_idx) = ty {
+                            let node_from = self.get_or_add_node(Node(def.function, *actual_idx));
+                            self.graph.add_edge(node_from, node_to, Edge::Identity);
+                        } else {
+                            let type_params = Self::extract_type_parameters(ty);
+                            for type_param in type_params {
+                                let node_from =
+                                    self.get_or_add_node(Node(def.function, type_param));
+                                self.graph.add_edge(node_from, node_to, Edge::TyConApp(&ty));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Computes the strongly connected components of the graph built and keep the ones that
+    /// contain at least one `TyConApp` edge. Such components indicate there exists a loop such
+    /// that an input type can get "bigger" infinitely many times along the path, also creating
+    /// infinitely many types. This is precisely the kind of constructs we want to forbid.
+    fn find_non_trivial_loops(&self) -> Vec<(Vec<NodeIndex>, Vec<EdgeIndex>)> {
+        tarjan_scc(&self.graph)
+            .into_iter()
+            .filter_map(move |nodes| {
+                let node_set: HashSet<_> = nodes.iter().cloned().collect();
+                let mut edges: Vec<EdgeIndex> = vec![];
+                let mut trivial = true;
+
+                for node_idx in &nodes {
+                    for edge in self.graph.edges(*node_idx) {
+                        if node_set.contains(&edge.target()) {
+                            edges.push(edge.id());
+                            match edge.weight() {
+                                Edge::Identity => (),
+                                Edge::TyConApp(_) => {
+                                    trivial = false;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                if trivial {
+                    None
+                } else {
+                    Some((nodes, edges))
+                }
+            })
+            .collect()
+    }
+
+    pub fn verify(mut self) -> Vec<VMStatus> {
+        self.build_graph();
+        let loops = self.find_non_trivial_loops();
+
+        loops
+            .into_iter()
+            .map(|(nodes, _edges)| {
+                VMStatus::new(StatusCode::LOOP_IN_INSTANTIATION_GRAPH).with_message(
+                    nodes
+                        .into_iter()
+                        .map(|idx| {
+                            let Node(func_handle_idx, type_param_idx) =
+                                self.graph.node_weight(idx).unwrap();
+                            format!(
```

Comment:
Looks like right now you are simply printing out all the nodes in each SCC that has a colored edge.  An SCC is not the same as a loop.

Reply:
That's a really good point. The reason why we care about loops is that if any one loop contains a colored edge, we have a problem. However one tricky issue: loops may share nodes or edges.
![Image](https://user-images.githubusercontent.com/10379359/65543659-6e7baa80-dec6-11e9-9d67-8a476ae0364f.png)
Even worse, the number of loops that contain a colored edge may grow exponentially:
![Image](https://user-images.githubusercontent.com/10379359/65543665-72a7c800-dec6-11e9-9f72-f67eb85d410b.png)
The algorithm I proposed is guaranteed to find the containing SCCs for all loops with colored edges, but it seems hard to me to remove the somewhat irrelevant edges & nodes (image 1) and unfeasible to report all loops due to exponential growth. 

Maybe there's sth we can do with the irrelevant edges & nodes, but for now I would suggest simply reporting the SCC in the error message.

Reply:
Or maybe we should simply report the colored edges? They are the ones that are really causing the trouble.

Reply:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.228.6872&rep=rep1&type=pdf

Look at Figure 2 in the paper above.  It describes an algorithm called "nested depth-first search".  This algorithm solves a problem very similar to what you are solving but it does not compute SCCs and also provides a witness cycle.  If you want, I can help you adapt the algorithm to your problem.  But you can also go ahead and land the diff just reporting the SCCs for now and revisit later at your leisure and pleasure.

Reply:
Thanks for bringing this up! I studied the paper and here's my thought:
1) The nested dfs (ndfs for short hereafter) algorithm finds a cycle that contains a colored node in a graph. This can be proven equivalent to the edge version easily. So this algorithm at least good enough for the on chain validator (yes/no).
2) The running time is O(size of graph), as each node is visited at most twice. The algorithm can terminate early once it finds one candidate. (Although I'd say this is also somewhat true for a manually implemented Tarjan's SCC.)
3) It seems to me that ndfs does not ensure all colored edges are reported. Here's an example:
![Image](https://user-images.githubusercontent.com/10379359/65639694-7ce8c580-df9d-11e9-9485-13414f399696.PNG)
Assume node 6 was reported as a cycle containing colored node 2. Then {6, 1} will be in the state space and when you visit {6, 1} from {5, 1} it is ignored. A modified algorithm may be able to say sth like 4 -> 5 -> 6 is attached to a cycle already found, but I feel like that's very similar to what Tarjan's SCC algo is doing.

Reply:
I agree.  NDFS is guaranteed to find one erroneous cycle.  It won't find all.